File: MatematicaApp/src/main/java/com/sergiocruz/Matematica/helper/MenuHelper.java
package com.sergiocruz.Matematica.helper;

import android.app.Activity;
import android.content.Intent;
import android.text.SpannableString;
import android.text.style.SuperscriptSpan;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.sergiocruz.Matematica.R;

import java.util.ArrayList;

import static com.sergiocruz.Matematica.R.id.history;
import static com.sergiocruz.Matematica.fragment.HomeFragment.mActivity;

/*****
 * Project Matematica
 * Package com.sergiocruz.Matematica.helper
 * Created by Sergio on 09/11/2016 21:11
 ******/

public class MenuHelper {

    private static ArrayList<View> getViewsByTag(ViewGroup root, String tag) {
        ArrayList<View> views = new ArrayList<View>();
        final int childCount = root.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = root.getChildAt(i);
            if (child instanceof ViewGroup) {
                views.addAll(getViewsByTag((ViewGroup) child, tag));
            }
            final Object tagObj = child.getTag(R.id.texto);
            if (tagObj != null && tagObj.equals(tag)) {
                views.add(child);
            }
        }
        return views;
    }

    public static void remove_history(Activity activity) {
        ViewGroup history = (ViewGroup) activity.findViewById(R.id.history);
        if ((history).getChildCount() > 0)
            (history).removeAllViews();
        Toast thetoast = Toast.makeText(activity, R.string.history_deleted, Toast.LENGTH_SHORT);
        thetoast.setGravity(Gravity.CENTER, 0, 0);
        thetoast.show();
    }

    public static void share_history(Activity activity) {
        ViewGroup history_view = (ViewGroup) activity.findViewById(history);
        ArrayList<View> textViews_withTAG_texto = getViewsByTag(history_view, "texto");
        if (textViews_withTAG_texto.size() > 0) {
            String text_fromTextViews_final = "";
            for (int i = 0; i < textViews_withTAG_texto.size(); i++) {
                if (textViews_withTAG_texto.get(i) instanceof TextView) {
                    String text_fromTextView = ((TextView) textViews_withTAG_texto.get(i)).getText().toString() + "\n";
                    SpannableString ss = new SpannableString(((TextView) textViews_withTAG_texto.get(i)).getText());
                    SuperscriptSpan[] spans = ss.getSpans(0, ((TextView) textViews_withTAG_texto.get(i)).getText().length(), SuperscriptSpan.class);
                    int corr = 0;
                    for (SuperscriptSpan span : spans) {
                        int start = ss.getSpanStart(span) + corr;
                        text_fromTextView = text_fromTextView.substring(0, start) + "^" + text_fromTextView.substring(start);
                        corr++;
                    }
                    text_fromTextViews_final += text_fromTextView;
                }
            }

            Intent sendIntent = new Intent();
            sendIntent.setAction(Intent.ACTION_SEND);
            sendIntent.putExtra(Intent.EXTRA_TEXT, activity.getResources().getString(R.string.app_long_description) +
                    activity.getResources().getString(R.string.app_version_name) + "\n" + text_fromTextViews_final);
            sendIntent.setType("text/plain");
            activity.startActivity(Intent.createChooser(sendIntent, mActivity.getResources().getString(R.string.app_name)));

        } else {
            Toast thetoast = Toast.makeText(activity, R.string.nothing_toshare, Toast.LENGTH_SHORT);
            thetoast.setGravity(Gravity.CENTER, 0, 0);
            thetoast.show();
        }
    }


}

File: MatematicaApp/src/main/java/com/sergiocruz/Matematica/helper/MenuHelper.kt
package com.sergiocruz.Matematica.helper

import android.content.Intent
import android.text.SpannableString
import android.text.style.SuperscriptSpan
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.sergiocruz.Matematica.R
import java.util.*

/*****
 * Project Matematica
 * Package com.sergiocruz.Matematica.helper
 * Created by Sergio on 09/11/2016 21:11
 */


private fun getViewsByTag(root: ViewGroup, tag: String?): ArrayList<View> {
    val views = ArrayList<View>()
    val childCount = root.childCount
    for (i in 0 until childCount) {
        val child = root.getChildAt(i)
        if (child is ViewGroup) views.addAll(getViewsByTag(child, tag))
        val tagObj = child.getTag(R.id.texto)
        if (tagObj != null && tagObj == tag) views.add(child)
    }
    return views
}

fun removeHistory(historyLayout: ViewGroup) {
    if (historyLayout.childCount > 0)
        historyLayout.removeAllViews()
    val theToast = Toast.makeText(historyLayout.context, R.string.history_deleted, Toast.LENGTH_SHORT)
    theToast.setGravity(Gravity.CENTER, 0, 0)
    theToast.show()
}

fun shareHistory(historyLayout: ViewGroup) {
    val textViewsWithTAGTexto = getViewsByTag(historyLayout, "texto")
    if (textViewsWithTAGTexto.size > 0) {
        var textFinal = ""
        for (i in textViewsWithTAGTexto.indices) {
            if (textViewsWithTAGTexto[i] is TextView) {
                var text = (textViewsWithTAGTexto[i] as TextView).text.toString() + "\n"
                val ss = SpannableString((textViewsWithTAGTexto[i] as TextView).text)
                val spans = ss.getSpans(0, (textViewsWithTAGTexto[i] as TextView).text.length, SuperscriptSpan::class.java)
                for ((corr, span) in spans.withIndex()) {
                    val start = ss.getSpanStart(span) + corr
                    text = text.substring(0, start) + "^" + text.substring(start)
                }
                textFinal += text
            }
        }

        val sendIntent = Intent()
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, historyLayout.context.resources.getString(R.string.app_long_description) +
                historyLayout.context.resources.getString(R.string.app_version_name) + "\n" + textFinal)
        sendIntent.type = "text/plain"
        historyLayout.context.startActivity(Intent.createChooser(sendIntent, historyLayout.context.resources.getString(R.string.app_name)))

    } else {
        val thetoast = Toast.makeText(historyLayout.context, R.string.nothing_toshare, Toast.LENGTH_SHORT)
        thetoast.setGravity(Gravity.CENTER, 0, 0)
        thetoast.show()
    }
}

fun biggerOrEqual(a: Int?, b: Int) = a?: b

fun nulableString() {
    val a: String? = null
    a?.plus("cenas")
}

fun areYouNullOrSomething(a: String?, b: String?) = a ?: b
@JvmOverloads
fun reformat(str: String,
             normalizeCase: Boolean? = true,
             upperCaseFirstLetter: Boolean = true,
             divideByCamelHumps: Boolean = false,
             wordSeparator: Char = ' ') {
}